core/services/relay/evm/mercury/queue.go

```diff
@@ -0,0 +1,166 @@
+package mercury
+
+import (
+	"container/heap"
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/services"
+	"github.com/smartcontractkit/chainlink/v2/core/services/relay/evm/mercury/wsrpc/pb"
+	ocrtypes "github.com/smartcontractkit/libocr/offchainreporting2/types"
+)
+
+var _ services.ServiceCtx = (*TransmitQueue)(nil)
+
+// TransmitQueue is the high-level package that everything outside of this file should be using
+// It stores pending transmissions, yielding the latest (highest priority) first to the caller
+type TransmitQueue struct {
+	cond sync.Cond
+	lggr logger.Logger
+	mu   *sync.RWMutex
+
+	pq     *priorityQueue
+	maxlen int
+	closed bool
+}
+
+type Transmission struct {
+	Req       *pb.TransmitRequest    // the payload to transmit
+	ReportCtx ocrtypes.ReportContext // contains priority information (latest epoch/round wins)
+
+	// The index is needed by update and is maintained by the heap.Interface
+	// methods
+	// It should NOT be set manually
+	index int // the index of the item in the heap
+}
+
+// maxlen controls how many items will be stored in the queue
+// 0 means unlimited - be careful, this can cause memory leaks
+func NewTransmitQueue(lggr logger.Logger, maxlen int) *TransmitQueue {
+	pq := new(priorityQueue)
+	heap.Init(pq) // for completeness
+	mu := new(sync.RWMutex)
+	return &TransmitQueue{sync.Cond{L: mu}, lggr.Named("TransmitQueue"), mu, pq, maxlen, false}
+}
+
+func (tq *TransmitQueue) Push(req *pb.TransmitRequest, reportCtx ocrtypes.ReportContext) (ok bool) {
+	tq.cond.L.Lock()
+	if tq.closed {
+		return false
```

Comment:
Unlock?
I think you could defer it beforehand, since the Signal docs say:
> It is allowed but not required for the caller to hold c.L during the call.